File: Part4_Splicing_Analysis/src/plot_anchored_gc.py
#!/usr/bin/env python3
"""
plot_anchored_gc.py

Standalone script to plot anchored GC content data (TSV) with customizable window size.
Usage:
    python3 src/plot_anchored_gc.py path/to/data.tsv --window 500 --output my_plot.png
"""

import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import argparse
import os

def main():
    parser = argparse.ArgumentParser(description="Plot Anchored Intron GC Content from TSV")
    parser.add_argument("input_tsv", help="Path to anchored_intron_gc_data.tsv")
    parser.add_argument("--window", type=int, default=1000, help="Plot window size (e.g., 500 for +/- 500bp). Max depends on generated data (usually 1000).")
    parser.add_argument("--output", default="custom_anchored_gc_plot.png", help="Output plot filename")
    parser.add_argument("--ymax", type=float, default=None, help="Optional Y-axis max limit")
    parser.add_argument("--ymin", type=float, default=None, help="Optional Y-axis min limit")
    
    args = parser.parse_args()
    
    if not os.path.exists(args.input_tsv):
        print(f"Error: File not found: {args.input_tsv}")
        return

    print(f"Loading {args.input_tsv}...")
    df = pd.read_csv(args.input_tsv, sep='\t')
    
    # Filter by window
    if args.window:
        print(f"Filtering data to +/- {args.window}bp...")
        df = df[(df['Position'] >= -args.window) & (df['Position'] <= args.window)]
        
    print("Plotting...")
    sns.set_theme(style="whitegrid")
    
    fig, axes = plt.subplots(1, 2, figsize=(12, 6))
    
    # Define custom palette if standard 'Lost'/'Preserved' group names are present
    palette = None
    groups = df['Group'].unique()
    if 'Lost' in groups and 'Preserved' in groups and 'Constitutive' in groups:
        palette = {'Lost': 'C1', 'Preserved': 'C0', 'Constitutive': 'C2'} # Adjust colors as needed
    
    # Start Codon
    sns.lineplot(data=df[df['Feature']=='Start_Codon'], x='Position', y='GC_Percent', hue='Group', palette=palette, ax=axes[0])
    axes[0].set_title(f"Start Codon Anchored GC Content (+/- {args.window}bp)")
    axes[0].set_ylabel("GC Fraction")
    if args.ymin is not None and args.ymax is not None:
        axes[0].set_ylim(args.ymin, args.ymax)
    
    # Stop Codon
    sns.lineplot(data=df[df['Feature']=='Stop_Codon'], x='Position', y='GC_Percent', hue='Group', palette=palette, ax=axes[1])
    axes[1].set_title(f"Stop Codon Anchored GC Content (+/- {args.window}bp)")
    axes[1].set_ylabel("GC Fraction")
    if args.ymin is not None and args.ymax is not None:
        axes[1].set_ylim(args.ymin, args.ymax)
    
    plt.tight_layout()
    plt.savefig(args.output)
    print(f"Saved plot to {args.output}")

if __name__ == "__main__":
    main()

File: Part4_Splicing_Analysis/src/plot_anchored_gc.R
#!/usr/bin/env Rscript

# plot_anchored_gc.R
# Usage: Rscript plot_anchored_gc.R <input_tsv> <window_size> <output_pdf>

args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 3) {
  stop("Usage: Rscript plot_anchored_gc.R <input_tsv> <window_size> <output_pdf>\nExample: Rscript plot_anchored_gc.R results/data.tsv 500 my_plot.pdf")
}

input_file <- args[1]
window_size <- as.numeric(args[2])
output_file <- args[3]

if (!file.exists(input_file)) {
  stop(paste("Input file not found:", input_file))
}

message("Loading libraries...")
suppressPackageStartupMessages({
  library(ggplot2)
  library(dplyr)
  library(readr)
  library(gridExtra)
})

message(paste("Reading", input_file))
df <- read_tsv(input_file, show_col_types = FALSE)

# Filter by window
message(paste("Filtering data to +/-", window_size, "bp"))
df_filtered <- df %>%
  filter(Position >= -window_size & Position <= window_size)

# Set up colors
# Match Python: Lost=C1 (Orange), Preserved=C0 (Blue), Constitutive=C2 (Green)
# R default palette is differnt, so let's set manual colors close to matplotlib's C0/C1/C2
custom_colors <- c(
  "Lost" = "#ff7f0e",       # C1 Orange
  "Preserved" = "#1f77b4",  # C0 Blue
  "Constitutive" = "#2ca02c", # C2 Green
  "UFM1_dependent" = "#ff7f0e",
  "UFM1_independent" = "#1f77b4"
)

message("Plotting...")

# Start Codon Plot
p_start <- ggplot(df_filtered %>% filter(Feature == "Start_Codon"), 
                  aes(x = Position, y = GC_Percent, color = Group)) +
  geom_line(size = 1) +
  scale_color_manual(values = custom_colors) +
  theme_minimal() +
  labs(title = paste("Start Codon Anchored GC Content (+/-", window_size, "bp)"),
       y = "GC Fraction", x = "Position relative to Start Codon") +
  theme(legend.position = "bottom")

# Stop Codon Plot
p_stop <- ggplot(df_filtered %>% filter(Feature == "Stop_Codon"), 
                 aes(x = Position, y = GC_Percent, color = Group)) +
  geom_line(size = 1) +
  scale_color_manual(values = custom_colors) +
  theme_minimal() +
  labs(title = paste("Stop Codon Anchored GC Content (+/-", window_size, "bp)"),
       y = "GC Fraction", x = "Position relative to Stop Codon") +
  theme(legend.position = "bottom")

# Combine and Save
message(paste("Saving to", output_file))
p_combined <- grid.arrange(p_start, p_stop, ncol = 2)
ggsave(output_file, p_combined, width = 12, height = 6)

message("Done.")
